addons_marketplace/marketplace-manager/services/marketplace_service.py

```diff
@@ -0,0 +1,38 @@
+import logging
+import threading
+
+import docker
+from db import marketplace_db
+
+
+def verify_addon(addon_id, addon):
+    logging.info(f"verifying addon-{addon_id}...")
+    client = docker.from_env()
+    for service in addon["services"]:
+        image_uri = service.get("image_uri")
+        try:
+            # TODO utilize opencontainers image spec to verify image instead of docker
+            logging.info(f"Pulling image: {image_uri}")
+            service = client.images.pull(image_uri)
+            logging.info(f"Image pulled: {image_uri}")
+
+            client.images.remove(service.id)
+        except docker.errors.ImageRemoveError as e:
+            logging.warning(f"Failed to remove image {image_uri}: {e}")
+        except docker.errors.DockerException as e:
+            logging.warning(f"Failed to pull image {image_uri}: {e}")
+            marketplace_db.update_addon(addon_id, {"status": "verification_failed"})
+            return
+
+    logging.info(f"Addon-{addon_id} verified")
+    marketplace_db.update_addon(addon_id, {"status": "approved"})
+
+
+def register_addon(addon):
+    addon = marketplace_db.create_addon({**addon, "status": "under_review"})
```

Comment:
What reason is there to store addons in the DB that failed the verification? Is is due to the long asynch behavior and letting callers wait for the response?
(I.e. why not just store addons in the DB that passed verification?)

Reply:
Currently I dont have a strong opinion about whether to keep it or remove failed ones. For the time being I dont see harm in keeping failed ones. Later if we find out that failed ones are no longer needed and they are taking a lot of space, we could setup a cron job the regularly cleans the db.